File: backend/app/services/outfits/slots.py
from __future__ import annotations

from enum import StrEnum


class OutfitSlot(StrEnum):
    TOP = "top"
    BOTTOM = "bottom"
    OUTER = "outer"
    SHOES = "shoes"


TOP_CATEGORIES = {"t-shirt", "shirt", "sweater"}
THIN_TOP_CATEGORIES = {"t-shirt", "shirt"}
BOTTOM_CATEGORIES = {"jeans", "pants", "shorts", "skirt"}
OUTER_CATEGORIES = {"hoodie", "jacket", "coat", "blazer"}
SHOES_CATEGORIES = {"sneakers", "shoes", "boots"}
DUAL_SLOT_CATEGORIES = {"dress"}


def category_to_slot(category: str | None) -> OutfitSlot | None:
    if not category:
        return None
    c = category.strip().lower()
    if c in TOP_CATEGORIES:
        return OutfitSlot.TOP
    if c in BOTTOM_CATEGORIES:
        return OutfitSlot.BOTTOM
    if c in OUTER_CATEGORIES:
        return OutfitSlot.OUTER
    if c in SHOES_CATEGORIES:
        return OutfitSlot.SHOES
    return None


def category_to_slots(category: str | None) -> set[OutfitSlot]:
    """Return the set of slots a category fills. Dresses fill both TOP and BOTTOM."""
    if not category:
        return set()
    if category.strip().lower() in DUAL_SLOT_CATEGORIES:
        return {OutfitSlot.TOP, OutfitSlot.BOTTOM}
    slot = category_to_slot(category)
    return {slot} if slot is not None else set()

File: android/app/src/main/java/com/example/outfitai/ui/outfits/Slots.kt
package com.example.outfitai.ui.outfits

// Mirrors backend/app/services/outfits/slots.py — keep in sync when categories change.
enum class Slot { TOP, BOTTOM, OUTER, SHOES }

private val TOP_CATEGORIES    = setOf("t-shirt", "shirt", "sweater")
private val BOTTOM_CATEGORIES = setOf("jeans", "pants", "shorts", "skirt")
private val OUTER_CATEGORIES  = setOf("hoodie", "jacket", "coat", "blazer")
private val SHOES_CATEGORIES  = setOf("sneakers", "shoes", "boots")

fun categoryToSlot(category: String?): Slot? {
    val c = category?.trim()?.lowercase() ?: return null
    return when {
        c in TOP_CATEGORIES    -> Slot.TOP
        c in BOTTOM_CATEGORIES -> Slot.BOTTOM
        c in OUTER_CATEGORIES  -> Slot.OUTER
        c in SHOES_CATEGORIES  -> Slot.SHOES
        else                   -> null
    }
}

fun slotCategories(slot: Slot): Set<String> = when (slot) {
    Slot.TOP    -> TOP_CATEGORIES
    Slot.BOTTOM -> BOTTOM_CATEGORIES
    Slot.OUTER  -> OUTER_CATEGORIES
    Slot.SHOES  -> SHOES_CATEGORIES
}
